### ISSUE TITLE
Support for marco-o1 model

### ISSUE THREAD

**Issue by spj2401Dev** [n4ze3m/page-assist#257]
Ollama just added a new model called [marco-o1](https://ollama.com/library/marco-o1):

This model has a <Thought></Thought> tag, where it thinks about what it should do, as its a reasoning model. 
The actual result is then delivered in the <Output> </Output> tag. 

Currently Page assist shows it like this:
<img width="562" alt="{A9CCB31C-FE05-4452-A62A-05156CF7D800}" src="https://github.com/user-attachments/assets/4c9f6387-5ec5-485c-a078-7ab7933ddc5d">

My Idea would be, to just deliver the Output by default, but then like the citations, I can see the thoughts of the AI model. A bit like this:
<img width="629" alt="{0D9CAA82-569E-46B0-A565-1C741F770682}" src="https://github.com/user-attachments/assets/edc249c8-0bfa-472c-b8e6-d5a3ea7d6bf0">

Then when you click on that, it should just show the Thought tag. (ofc properly formated then)

<img width="612" alt="{574E6CBD-CD8F-4450-823E-FAB2DF84B026}" src="https://github.com/user-attachments/assets/66eb5304-7604-430a-8e2e-dc0c769d910c">

**Comment by n4ze3m** [comment]
Yes, this one is already planned. I'll add it in the next release. :)

**Comment by nonetrix** [comment]
Please make it generalized somewhat because I'd imagine other models might have their own way of doing this, if anything I hope this is abstracted by Ollama eventually

**Comment by ucffool** [comment]
It's a qwen model so use this System Prompt:
```
{{- if .Suffix }}<|fim_prefix|>{{ .Prompt }}<|fim_suffix|>{{ .Suffix }}<|fim_middle|>
{{- else if .Messages }}
{{- if or .System .Tools }}<|im_start|>system
{{- if .System }}
{{ .System }}
{{- end }}
{{- if .Tools }}

# Tools

You may call one or more functions to assist with the user query.

You are provided with function signatures within <tools></tools> XML tags:
<tools>
{{- range .Tools }}
{"type": "function", "function": {{ .Function }}}
{{- end }}
</tools>

For each function call, return a json object with function name and arguments within <tool_call></tool_call> XML tags:
<tool_call>
{"name": <function-name>, "arguments": <args-json-object>}
</tool_call>
{{- end }}<|im_end|>
{{ end }}
{{- range $i, $_ := .Messages }}
{{- $last := eq (len (slice $.Messages $i)) 1 -}}
{{- if eq .Role "user" }}<|im_start|>user
{{ .Content }}<|im_end|>
{{ else if eq .Role "assistant" }}<|im_start|>assistant
{{ if .Content }}{{ .Content }}
{{- else if .ToolCalls }}<tool_call>
{{ range .ToolCalls }}{"name": "{{ .Function.Name }}", "arguments": {{ .Function.Arguments }}}
{{ end }}</tool_call>
{{- end }}{{ if not $last }}<|im_end|>
{{ end }}
{{- else if eq .Role "tool" }}<|im_start|>user
<tool_response>
{{ .Content }}
</tool_response><|im_end|>
{{ end }}
{{- if and (ne .Role "assistant") $last }}<|im_start|>assistant
{{ end }}
{{- end }}
{{- else }}
{{- if .System }}<|im_start|>system
{{ .System }}<|im_end|>
{{ end }}{{ if .Prompt }}<|im_start|>user
{{ .Prompt }}<|im_end|>
{{ end }}<|im_start|>assistant
{{ end }}{{ .Response }}{{ if .Response }}<|im_end|>{{ end }}```

**Comment by Laserology** [comment]
This issue should be renamed to "Add support for COT (Thinking) models.
Deepseek's R1 release brings COT models to the masses (8b and 32b models can run OK on a lot of hardware), and proper support for the <think> tag is needed.

As for generalization, not sure? Support for others would definitely be welcome but i think the most common tag is <think> from what i have seen personally so far. Perhaps a dictionary of think tags based on model name that gets hard coded.

**Comment by Laserology** [comment]
> Hey, It will be available in this weekend's update. :)
> 
> ![image](https://github.com/user-attachments/assets/ce5335cf-5b42-4c49-b08b-c3d447cda040)
> 

Great! Thanks a bunch.
Your work is much appreciated.

**Comment by spj2401Dev** [comment]
Amazing, thank you so much! This improves page assist so much

**Comment by nonetrix** [comment]
I assume this can be closed?

**Comment by n4ze3m** [comment]
Yes, I was waiting for the Chrome Web Store to approve the new release.
